### ISSUE TITLE
Cannot use ResultSetMetaData.getPrecision() to determine columns size.

### ISSUE THREAD

**Issue by vlad-lyutenko** [trinodb/trino#15950]
During work on string predicate pushdown fro case sensitive columns (https://github.com/trinodb/trino/pull/15714).

I tried to switch from 
 `public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, SchemaTableName schemaTableName)`
to 
`public JdbcTableHandle getTableHandle(ConnectorSession session, PreparedQuery preparedQuery)`

So I tried to get table handle/ column handles not from DatabaseMetada, but from ResultSetMetada (for select * query).

But then I hit such issue, that when we determine column size in first case we use : 
`getInteger(resultSet, "COLUMN_SIZE"),`

But in second case we are using : 
`metadata.getPrecision(column)`

But according to documentation of getPrecision:
` For character data, this is the length in characters. `

So instead of size in bytes we can get size in characters. And basically I hit this issue in one of the tests:
where we are creating varchar(max) column, with size - 2147483647 (the same showed in DBeaver) but getPrecision() returns 1073741823, like twice less (so it's assumed 2 bytes encoding per character as far as I undestand)

**Comment by kokosing** [comment]
What is the user experience visible outcome of the problem?  Are you saying that we have bug in type mapping? How does it manifest?

**Comment by vlad-lyutenko** [comment]
The case I see : 
In database column has type Varchar(2147483647) - In trino Varchar(1073741823).
How it hits the user, I am not sure.

**Comment by vlad-lyutenko** [comment]
I think I forgot to mention that this problem will be present for queries like this:
```
SELECT * FROM TABLE(
                       sqlserver.system.query(
                           query => 'SELECT * FROM lineitem'
                       )
                    );
```
Because only for nested query, we are using public 
`JdbcTableHandle getTableHandle(ConnectorSession session, PreparedQuery preparedQuery)`

For other cases use
`public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, SchemaTableName schemaTableName)`
Where we correctly resolve column size, using DatabaseMetadata.

**Comment by kokosing** [comment]
Are you saying that:

```
DESCRIBE SELECT * FROM TABLE(sqlserver.system.query(query => 'SELECT * FROM lineitem'));
```
will return different data than
```
DESCRIBE SELECT * FROM sqlserver.schema.lineitem;
```
?

If we would use https://github.com/trinodb/trino/pull/15943, would it help to capture the same information for type mapping disregarding how we access table?

CC: @kasiafi @findepi

**Comment by vlad-lyutenko** [comment]
Yes, in first case column size could be differ from second one (correct one)

**Comment by findepi** [comment]
> Yes, in first case column size could be differ from second one (correct one)

Not sure if this is a problem though.

**Comment by vlad-lyutenko** [comment]
> If we would use #15943, would it help to capture the same information for type mapping disregarding how we access table?

In this case you can get incorrect column size https://github.com/trinodb/trino/issues/15950#issuecomment-1414114213

**Comment by kokosing** [comment]
> In database column has type Varchar(2147483647) - In trino Varchar(1073741823).

The numbers are different, but maybe their semantics is different too, notice that these `varchar` types comes from to different databases.

See SqlServer docs: https://learn.microsoft.com/en-us/sql/t-sql/data-types/char-and-varchar-transact-sql?view=sql-server-ver16#varchar---n--max--

> Use n to define the string size in bytes 

Trino docs: https://trino.io/docs/current/language/types.html#varchar

> Variable length character data with an optional maximum length.

So it is ok that these numbers are different.

**Comment by vlad-lyutenko** [comment]
> So it is ok that these numbers are different.

Ok it could be different between database and Trino.
But it's also different in Trino itself, depending how we retrieve data.  

https://github.com/trinodb/trino/issues/15950#issuecomment-1415792298
